demos/digest/EVP_MD_demo.c

```diff
@@ -0,0 +1,189 @@
+	"And lose the name of Action. Soft you now,\n"
+	"The fair Ophelia? Nymph in thy Orisons\n"
+	"Be all my sins remember'd.\n"
+; 
+
+/*
+ * The known value of the SHA3-512 digest of the above soliloqy
+ */
+const unsigned char known_answer[] = {
+	0xbb, 0x69, 0xf8, 0x09, 0x9c, 0x2e, 0x00, 0x3d,
+	0xa4, 0x29, 0x5f, 0x59, 0x4b, 0x89, 0xe4, 0xd9,
+	0xdb, 0xa2, 0xe5, 0xaf, 0xa5, 0x87, 0x73, 0x9d,
+	0x83, 0x72, 0xcf, 0xea, 0x84, 0x66, 0xc1, 0xf9,
+	0xc9, 0x78, 0xef, 0xba, 0x3d, 0xe9, 0xc1, 0xff,
+	0xa3, 0x75, 0xc7, 0x58, 0x74, 0x8e, 0x9c, 0x1d,
+	0x14, 0xd9, 0xdd, 0xd1, 0xfd, 0x24, 0x30, 0xd6,
+	0x81, 0xca, 0x8f, 0x78, 0x29, 0x19, 0x9a, 0xfe,
+};
+
+int demonstrate_digest(void) {
+	OSSL_LIB_CTX	*library_context;
+	int result = 0;
+
+	library_context = OSSL_LIB_CTX_new();
+	if (library_context == NULL) {
+		fprintf(stderr, "OSSL_LIB_CTX_new() returned NULL\n");
+		ERR_print_errors_fp(stderr);
+		return 0;
+	}
+
+	/*
+	 * Fetch a message digest by name
+	 */
+	const char * digest_name = "SHA3-512";
+	const char * option_properties = NULL;
+	EVP_MD *message_digest = EVP_MD_fetch(library_context,
+								digest_name, option_properties);
+	if (message_digest == NULL) {
+		fprintf(stderr, "EVP_MD_fetch could not find %s.", digest_name);
+        ERR_print_errors_fp(stderr);
+		OSSL_LIB_CTX_free(library_context);
+		return 0;
+	}
+/*
+ * Determine the length of the fetched digest type
+ */
+	unsigned int digest_length = EVP_MD_size(message_digest);
+	if (digest_length <= 0) {
+		fprintf(stderr, "EVP_MD_size returned invalid size.\n");
+		goto cleanup_md;
+	}
+
+	unsigned char *digest_value = malloc(digest_length);
+	if (digest_value == NULL) {
+		fprintf(stderr, "No memory.\n");
+		goto cleanup_md;
+	}
+/*
+ * Make a message digest context to hold temporary state
+ * during digest creation
+ */
+	EVP_MD_CTX *digest_context = EVP_MD_CTX_new();
+	if (digest_context == NULL) {
+		fprintf(stderr, "EVP_MD_CTX_new failed.\n");
+        ERR_print_errors_fp(stderr);
+		goto cleanup_md;
+	}
+/*
+ * Initialize the message digest context to use the fetched 
+ * digest provider
+ */
+	if (EVP_DigestInit(digest_context, message_digest) != 1) {
+		fprintf(stderr, "EVP_DigestInit failed.\n");
+        ERR_print_errors_fp(stderr);
+		goto cleanup_md_ctx;
+	}
+/*
+ * Digest parts one and two of the soliloqy
+ */
+	if (EVP_DigestUpdate(digest_context, hamlet_1, strlen(hamlet_1)) != 1) {
+		fprintf(stderr, "EVP_DigestUpdate(hamlet_1) failed.\n");
+        ERR_print_errors_fp(stderr);
+		goto cleanup_md_ctx;
+	}
+	if (EVP_DigestUpdate(digest_context, hamlet_2, strlen(hamlet_2)) != 1) {
+		fprintf(stderr, "EVP_DigestUpdate(hamlet_2) failed.\n");
+        ERR_print_errors_fp(stderr);
+		goto cleanup_md_ctx;
+	}
+	if (EVP_DigestFinal(digest_context, digest_value, &digest_length) != 1) {
+		fprintf(stderr, "EVP_DigestFinal() failed.\n");
+        ERR_print_errors_fp(stderr);
+		goto cleanup_md_ctx;
+	}
+	for( int ii=0; ii<digest_length; ii++ ) {
```

Comment:
Should look like this (j must be declared at the top)
for (j = 0; j < digest_length; j++) {